File: java/src/main/java/com/pqc/mlkem/Params.java
package com.pqc.mlkem;

/**
 * ML-KEM parameter sets (FIPS 203 Table 2).
 */
public final class Params {

    public final String name;
    public final int k;
    public final int eta1;
    public final int eta2;
    public final int du;
    public final int dv;
    public final int ekSize;
    public final int dkSize;
    public final int ctSize;

    public Params(String name, int k, int eta1, int eta2, int du, int dv) {
        this.name = name;
        this.k = k;
        this.eta1 = eta1;
        this.eta2 = eta2;
        this.du = du;
        this.dv = dv;
        // ekSize = 384*k + 32
        this.ekSize = 384 * k + 32;
        // dkSize = 768*k + 96
        this.dkSize = 768 * k + 96;
        // ctSize = 32*(du*k + dv)
        this.ctSize = 32 * (du * k + dv);
    }

    public static final Params ML_KEM_512 = new Params("ML-KEM-512", 2, 3, 2, 10, 4);
    public static final Params ML_KEM_768 = new Params("ML-KEM-768", 3, 2, 2, 10, 4);
    public static final Params ML_KEM_1024 = new Params("ML-KEM-1024", 4, 2, 2, 11, 5);
}

File: python/mlkem/params.py
"""ML-KEM parameter sets per FIPS 203."""

from dataclasses import dataclass


@dataclass(frozen=True)
class MLKEMParams:
    name: str
    k: int
    eta1: int
    eta2: int
    du: int
    dv: int

    @property
    def ek_size(self) -> int:
        return 384 * self.k + 32

    @property
    def dk_size(self) -> int:
        return 768 * self.k + 96

    @property
    def ct_size(self) -> int:
        return 32 * (self.du * self.k + self.dv)


ML_KEM_512 = MLKEMParams(name="ML-KEM-512", k=2, eta1=3, eta2=2, du=10, dv=4)
ML_KEM_768 = MLKEMParams(name="ML-KEM-768", k=3, eta1=2, eta2=2, du=10, dv=4)
ML_KEM_1024 = MLKEMParams(name="ML-KEM-1024", k=4, eta1=2, eta2=2, du=11, dv=5)
